src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -150,6 +150,209 @@ CosineTree::CosineTree(const arma::mat& dataset,
   ConstructBasis(treeQueue);
 }
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+  //dataset((other.parent == NULL) ? other.dataset : NULL)
+  dataset(other.dataset),
+  delta(other.delta),
+  parent(other.Parent()),
+  left(other.Left()),
+  right(other.Right()),
+  indices(other.indices),
+  l2NormsSquared(other.l2NormsSquared),
+  centroid(other.centroid),
+  basisVector(other.basisVector),
+  splitPointIndex(other.SplitPointIndex()),
+  numColumns(other.NumColumns()),
+  l2Error(other.L2Error()),
+  frobNormSquared(other.FrobNormSquared())
+{
+  std::vector<CosineTree*> nodeStack;
+  nodeStack.push_back(this);
+
+  // While stack is not empty.
+  while (nodeStack.size())
+  {
+    // Pop a node from the stack and split it.
+    CosineTree *currentNode, *currentLeft, *currentRight;
+    currentNode = nodeStack.back();
+    currentNode->CosineNodeSplit();
+    nodeStack.pop_back();
+
+    // Obtain pointers to the children of the node.
+    currentLeft = currentNode->Left();
+    currentRight = currentNode->Right();
+
+    // If children exist.
+    if (currentLeft && currentRight)
+    {
+      // Push the child nodes on to the stack.
+      nodeStack.push_back(currentLeft);
+      nodeStack.push_back(currentRight);
+
+      // Obtain the split point of the popped node.
+      arma::vec splitPoint = data.col(currentNode->SplitPointIndex());
+
+      // Column indices of the the child nodes.
+      std::vector<size_t> leftIndices, rightIndices;
+      leftIndices = currentLeft->VectorIndices();
+      rightIndices = currentRight->VectorIndices();
+
+      // Calculate the cosine values for each of the columns in the node.
+      arma::vec cosines;
+      cosines.zeros(currentNode->NumColumns());
+
+      size_t i, j, k;
+      for (i = 0; i < leftIndices.size(); i++)
+        cosines(i) = arma::norm_dot(data.col(leftIndices[i]), splitPoint);
+
+      for (j = 0, k = i; j < rightIndices.size(); j++, k++)
+        cosines(k) = arma::norm_dot(data.col(rightIndices[j]), splitPoint);
+    }
+  }
+}
+
+//! Copy Assignment
+CosineTree&
+CosineTree::
+operator=(const CosineTree& other)
+{
+  // Return if it's the same tree.
+  if (this == &other)
+    return *this;
+
+  dataset = (other.parent == NULL) ? other.dataset : NULL;
```

Comment:
@sriramsk1999 this one does still need to be fixed. :+1:  It turns out, it has some other implications too: `CosineTree` will now need to track whether it is responsible for deleting the `dataset` pointer when it is deleted.  I didn't realize that earlier, sorry about that.  It will take just a little bit of refactoring but shouldn't be too hard I think. :)

Reply:
This comment also has not been addressed.